File: 1. Learn the Basics/2. Basics Of Maths/C++/5. Print All Divisors.cpp
/*
Print all divisors of a given number
*/

#include <bits/stdc++.h>
using namespace std;

void printDivisors(int n){
    vector<int> list;
    //O(sqrt(N)) --> N = number
    for(int i=1;i*i<=n;i++){
        if(n%i==0){
            list.push_back(i);
            if(n/i!=i){
                list.push_back(n/i);
            }
        }

    }
    //O(nlogn) --> n = number of divisors
    sort(list.begin(),list.end());
    //O(n) -->  n = number of divisors
    for(auto l : list){
        cout << l << " ";
    }
}

int main() {
	// your code goes here
    int n;
	cin >> n;
	printDivisors(n); //Total time complexity = O(N) + O(nlogn) + O(n)
	return 0;
}

File: 1. Learn the Basics/2. Basics Of Maths/Java/5. Print All Divisors.java
/* package codechef; // don't place package name! */

import java.util.*;
import java.lang.*;
import java.io.*;

/* Name of the class has to be "Main" only if the class is public. */

/*
Print all divisors of a given number
*/

class Codechef
{
    public static void printDivisors(int n){
        List<Integer> list = new ArrayList<>();
        for(int i = 1;i*i<=n;i++){
            if(n%i==0){
                list.add(i);
                if(n/i!=i){
                    list.add(n/i);
                }
            }
        }
        Collections.sort(list);
        for(int l : list){
            System.out.print(l+" ");
        }
    }
	public static void main (String[] args) throws java.lang.Exception
	{
		// your code goes here
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		printDivisors(n);
	}
}
